ports/rp2/rp2_pio.c

```diff
@@ -742,13 +742,44 @@ STATIC mp_obj_t rp2_state_machine_irq(size_t n_args, const mp_obj_t *pos_args, m
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_KW(rp2_state_machine_irq_obj, 1, rp2_state_machine_irq);
 
+// StateMachine.restart()
+STATIC mp_obj_t rp2_state_machine_restart(mp_obj_t self_in, mp_obj_t prog_in) {
```

Comment:
IMO this is very inconvenient to need to pass in the program.  If you had the program you could almost do `sm.init(prog)` to restart it (but not exactly that because you still need to specify the other arguments like `freq` etc).

If all it needs is to know the initial PC, this could be saved/cached in the `rp2_state_machine_obj_t` instance, then no prog needs to be passed in here.

Reply:
I did not like too passing in the program. sm.init() seemed to slow and bulky. The suggestion of  caching the initial PC in `rp2_state_machine_obj_t` sounds promising, so I'll try that,

Reply:
if my old eyes do not deceive me, the table of state machines with rp2_state_machine_obj_t objects is a const which cannot be changed. And the instance will again point to that const object. Instead, using a small table for starting addresses, indexed my the id of the state machine, does the job.

Reply:
Indeed!